Enable thp(transparent huge pages) for buffer sizes >=2MB

snadampal opened the pull request:
The 2MB thp pages provide better allocation latencies compared to the standard 4KB pages. This change has shown significant improvement for batch mode usecases where the tensor sizes are larger than 100MB.

Only enabled if `THP_MEM_ALLOC_ENABLE` environment variable is set.


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

ezyang commented:
This seems reasonable. Also cc'ing our Intel friends

ezyang commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -81,6 +88,15 @@ void* alloc_cpu(size_t nbytes) {
+      int ret = madvise(data, nbytes, MADV_HUGEPAGE);
+      if (ret != 0) {
+#ifdef DEBUG
+          fprintf(stderr, "thp madvise failed with %s\n", strerror(errno));
```
Hmm, is spewing to stderr in debug mode really the right way to do this :/ Maybe we should do a `TORCH_WARN_ONCE` here

snadampal replied:
thanks @ezyang , I've added TORCH_WARN_ONCE and also addressed the linter warning.

ezyang commented:
I think this will affect our Caffe2 workflows. Also cc'ing @swolchok for a look.

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -70,6 +70,31 @@ void* alloc_cpu(size_t nbytes) {
       "DefaultCPUAllocator: not enough memory: you tried to allocate ",
       nbytes,
       " bytes.");
+#elif defined(__linux__)
```
Hmm, should this change be limited to only x86 Linuxes? (I.e. I'm not sure if HugeTLB is enabled on ARM by default) 
Also, wouldn't one need to check that they are enabled by checking value of `/sys/kernel/mm/transparent_hugepage/enabled`?
```suggestion
#elif defined(__linux__) && defined(__x86_64__)
```

snadampal replied:
Hi @malfet , In fact I've primarily tested this feature on aarch64, the THP and HugeTLBs are fully supported on aarch64. 
And regarding checking the sys config, as you know the valid values for 
`/sys/kernel/mm/transparent_hugepage/enabled` are `always`, `madvise`, `never`

For most of the distros the default value is `madvise` for which we need to explicitly set the madvise(,,MADV_HUGEPAGE). that's what the patch does.
I have also tested this feature with 'always' and 'never' settings and the feature works as expected: 
for 'always' THP would be always enabled, and the explicit madvise() is redundant, but doesn't throw any error
for 'never' this memory advise hint is ignored.

jgong5 commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -70,6 +70,31 @@ void* alloc_cpu(size_t nbytes) {
+      strerror(err),
+      ")");
+  // enable thp (transparent huge pages) for larger buffers
+  if (nbytes >= gAlloc_threshold_thp) {
+    int ret = madvise(data, nbytes, MADV_HUGEPAGE);
+    if (ret != 0) {
+      TORCH_WARN_ONCE(
+          "thp madvise for HUGEPAGE failed with %s\n", strerror(errno));
+    }
```
Can we make it as a user opt-in option to enable THP on large memory allocation? THP reduces TLB miss rate but might have side effects like larger memory consumption, more fragmentations, kernel defragmentation overhead, perf fluctuation of kernel clear_page. The benefit would be workload specific and with tradeoffs.

snadampal replied:
one option to disable the thp system wide is by setting
`echo never > /sys/kernel/mm/transparent_hugepage/enabled`
If there is any runtime prop based feature enable/disable in pytorch please let me know, i will try to use that for selectively disabling it.

jgong5 replied:
I was referring to what is the default behavior without additional user settings. For the system-wide setting, `madvise` is the default. For PyTorch, I suggested we make this an opt-in, not enabled by default. One option is to have a frontend API for users to enable explicitly.

snadampal replied:
Hi @jgong5 , I agree with having configurability a better option, but I'm thinking how well user be made aware of it.
if you could point me to any other feature currently being controlled via frontend API, I would like to understand how they get used today. thank you!

jgong5 replied:
Currently, there are some control flags under `torch.backends.xyz` as well as other places for example, e.g.,
`torch.backends.mkldnn.*`: https://github.com/pytorch/pytorch/blob/11f51e798f480680a7634fbee2c3fb3cea09f970/torch/backends/mkldnn/__init__.py#L75, and `torch.backends.cudnn.*` etc.

My two cents on the places to add the flag. One option is to add it under `torch.cpu`, e.g., `torch.cpu.mm.transparent_hugepage.enabled`.

snadampal replied:
thank you! will take a look.

malfet commented:
/easycla

snadampal commented on c10/core/alignment.h:
```diff
@@ -14,4 +17,9 @@ constexpr size_t gAlignment = 16;
 
+#ifdef __linux__
+// inorder to enable thp, buffers need to be page aligned
+const size_t gAlignment_thp = sysconf(_SC_PAGESIZE);
```
sure.

snadampal commented:
> /easycla

yes, @malfet , I'm working with our org easyCLA manager, it's in progress.

snadampal commented:
Hi @jgong5 , I've added the error handling for sysconf and frontend api to control the c10 thp feature, added `torch.cpu.mm.transparent_hugepages.enabled`.

This completes all the comments so far. 
One thing I wanted to highlight is , to implement the frontend API support, I had to make c10 library dependent on torch, for the `at::globalContext` which I feel is not very elegant from the design point of view. But I didn't find any alternative to avoid the aten context dependency. Please let me know what you think.

Hi @malfet , our CLA is in the final stage approval, will provide the update on Monday.

jgong5 commented:
> One thing I wanted to highlight is , to implement the frontend API support, I had to make c10 library dependent on torch, for the `at::globalContext` which I feel is not very elegant from the design point of view. But I didn't find any alternative to avoid the aten context dependency. Please let me know what you think.

Agree that it is not good to have c10 depending on torch. I don't have a good idea either. @ezyang any suggestions?

snadampal commented:
the CI failures are mostly around the new torch dependency, the fix is going to be similar to what I've done for the c10 and caffe2 tests in the PR. I would like to hear from pytorch team on the c10->torch dependency I mentioned above before I go and update these tests.

```
-  target_link_libraries(${test_name}_${CPU_CAPABILITY} c10 sleef gtest_main)
+  target_link_libraries(${test_name}_${CPU_CAPABILITY} c10 torch sleef gtest_main)
```

btw, I've added the torch api support as a separate commit, if we are comfortable with enabling the c10 thp support by default without the torch api, we can go ahead with the first commit alone from this PR which doesn't introduce any unnecessary dependencies between the libraries.

ezyang commented:
Cannot have c10 depend on torch. Need to figure out some othe rway to do it.

snadampal commented:
Yeah, that's why I mentioned explicitly. I think this is going to be the issue if any feature in c10 to be controlled via torch api.  Other option I see is controlling it through some env variable, similar to most of the DNNL variables like ONEDNN_PRIMITIVE_CACHE_CAPACITY  or LRU_CACHE_CAPACITY etc...
what do you say?

snadampal commented:
Hi @jgong5 , @ezyang , I've added env variable based control for c10 thp feature and removed the torch api support. there is no more c10->torch dependency now.

jgong5 reviewed (approved):
LGTM but needs @ezyang to stamp on it too. Also, since it is a user-facing thing, do we need a doc? cc @malfet

snadampal commented:
Hi @malfet , @ezyang , easyCLA is done, and @jgong5 has approved the changes so far. Could one of you please let me know if I need to update any other docs/files for this feature. Thank you!

malfet commented:
@pytorchbot rebase

ezyang commented:
@malfet said he's going to do the final review for this

snadampal commented:
hi @malfet, wondering if there is anything pending on me for getting this PR merged. please let me know. thank you!

atalman commented:
Hi @malfet could you please take a look at this issue ? I

malfet commented on c10/core/impl/alloc_cpu.h:
```diff
@@ -4,8 +4,21 @@
 namespace c10 {
 
+#ifdef __linux__
+// since the default thp pagesize is 2MB, enable thp only
+// for buffers of size 2MB or larger to avoid memory bloating
+constexpr size_t gAlloc_threshold_thp = 2 * 1024 * 1024;
```
This constexpr seems to belong in alignment.h.
And why should it be guarded under `#ifdef __linux__` as transparent huge pages are hardware independent?

snadampal replied:
we are not enabling this feature on other OSes, to avoid unused variable warnings, i'm defining them under __linux__ only

malfet commented on c10/core/impl/alloc_cpu.h:
```diff
@@ -4,8 +4,21 @@
+// since the default thp pagesize is 2MB, enable thp only
+// for buffers of size 2MB or larger to avoid memory bloating
+constexpr size_t gAlloc_threshold_thp = 2 * 1024 * 1024;
+static const char* thp_env = std::getenv("C10_THP_MEM_ALLOC_ENABLE");
+const bool gIs_c10_thp_mem_alloc_enabled = std::atoi(thp_env != nullptr ? thp_env : "0");
```
This initializer is called at library load time, but if it were to be implemented as an inline function in alloc_cpu.cpp it could be postponed until the first use time

Something like(also, c10 name is redundant as it is already in c10 namespace):
```c++
const bool is_thp_alloc_enabled() {
  static bool value = [] {
      auto thp_env = std::getenv("C10_THP_MEM_ALLOC_ENABLE");
      return std::atoi(thp_env != nullptr ? thp_env : "0");
  };
  return value;
}

snadampal replied:
My intention was to avoid reading the env for every alloca call.

snadampal commented:
Thanks for the feedback, @malfet.  I will fix them.

snadampal commented:
Hi @malfet , I've incorporated most of your feedback. Please check and let me know.

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -41,6 +41,38 @@ void memset_junk(void* data, size_t num) {
 
+static inline bool is_thp_alloc_enabled() {
+  static bool value = [&](const char* pt) {
+    if (pt != NULL) {
```
nit
```suggestion
    if (pt != nullptr) {
```

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -81,6 +113,17 @@ void* alloc_cpu(size_t nbytes) {
+    int ret = madvise(data, nbytes, MADV_HUGEPAGE);
+    if (ret != 0) {
+      TORCH_WARN_ONCE(
+          "thp madvise for HUGEPAGE failed with %s\n", strerror(errno));
```
TORCH_WARN_ONCE does not take format strings, but rather concatenates output
```suggestion
          "thp madvise for HUGEPAGE failed with ", strerror(errno));
```

malfet commented on c10/core/impl/alloc_cpu.h:
```diff
@@ -4,6 +4,12 @@
 
 #include <cstddef>
 
+#ifdef __linux__
+#include <sys/mman.h>
+#include <unistd.h>
+#include <cstdlib>
+#endif
+
```
Those are not needed in `alloc_cpu.h`, are they?
```suggestion
```

snadampal replied:
#include <sys/mman.h>
#include <unistd.h>
are required for _SC_PAGESIZE and MADV_HUGEPAGE definitions.

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -41,6 +41,38 @@ void memset_junk(void* data, size_t num) {
+  return gAlignment;
+}
+
+constexpr bool is_thp_alloc(size_t nbytes) {
```
Aren't this one needed to suppress compilation warnings on say MacOS?
```suggestion
constexpr bool is_thp_alloc(C10_UNUSED size_t nbytes) {
```

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -41,6 +41,38 @@ void memset_junk(void* data, size_t num) {
+    } else {
+      return 0;
+    }
+  }(std::getenv("THP_MEM_ALLOC_ENABLE"));
```
This indeed works fine, but wouldn't it be nicer to just move this into the lambda body?
```suggestion
  }(std::getenv("THP_MEM_ALLOC_ENABLE"));
```

malfet reviewed (approved):
Looks good to me, but please address some feedback (i.e. use `C10_UNUSED`, do not use `NULL` as c++11 provides nice `nullptr`)

snadampal committed:
Enable thp(transparent huge pages) for buffer sizes >=2MB

The 2MB thp pages provide better allocation latencies compared
to the standard 4KB pages. This change has shown substantial
improvement for batch mode usecases where the tensor sizes
are larger than 100MB.

snadampal commented:
The CI reported one functional failure for CUDA, but I'm not sure whether the CUDA test failure is related to this PR. 

> test_ops.py::TestCommonCUDA::test_out_warning_linalg_lu_factor_cuda - AssertionError: Tensor-likes are not close!

malfet commented:
@pytorchbot merge

malfet commented on c10/core/impl/alloc_cpu.cpp:
```diff
@@ -41,6 +41,38 @@ void memset_junk(void* data, size_t num) {
+
+#ifdef __linux__
+inline size_t c10_compute_alignment(size_t nbytes) {
+  static int pagesize = sysconf(_SC_PAGESIZE);
```
```suggestion
  static const auto pagesize = sysconf(_SC_PAGESIZE);
```

malfet committed:
Update c10/core/impl/alloc_cpu.cpp

malfet commented:
@pytorchbot merge

seemethere commented:
@pytorchbot revert -c ghfirst -m "Reverting due to internal build issues, Meta employees see: https://fburl.com/sandcastle/1p4zvldk"

seemethere commented:
Would suggest applying the following patch to re-land:

```diff
diff --git a/c10/core/impl/alloc_cpu.cpp b/c10/core/impl/alloc_cpu.cpp
index 644f35f8de0..59b9d21b7d9 100644
--- a/c10/core/impl/alloc_cpu.cpp
+++ b/c10/core/impl/alloc_cpu.cpp
@@ -53,23 +53,23 @@ static inline bool is_thp_alloc_enabled() {
 }
 
 #ifdef __linux__
-inline size_t c10_compute_alignment(size_t nbytes) {
+C10_UNUSED inline size_t c10_compute_alignment(size_t nbytes) {
   static const auto pagesize = sysconf(_SC_PAGESIZE);
   // for kernels that don't provide page size, default it to 4K
   const size_t thp_alignment = (gPagesize < 0 ? gPagesize : pagesize);
   return (is_thp_alloc_enabled() ? thp_alignment : gAlignment);
 }
 
-inline bool is_thp_alloc(size_t nbytes) {
+C10_UNUSED inline bool is_thp_alloc(size_t nbytes) {
   // enable thp (transparent huge pages) for larger buffers
   return (is_thp_alloc_enabled() && (nbytes >= gAlloc_threshold_thp));
 }
 #else
-constexpr size_t c10_compute_alignment(C10_UNUSED size_t nbytes) {
+C10_UNUSED constexpr size_t c10_compute_alignment(C10_UNUSED size_t nbytes) {
   return gAlignment;
 }
 
-constexpr bool is_thp_alloc(C10_UNUSED size_t nbytes) {
+C10_UNUSED constexpr bool is_thp_alloc(C10_UNUSED size_t nbytes) {
   return false;
 }
 #endif

```

malfet commented:
> @malfet maybe it'd be good for us to add `-Werror,-Wunused-function` to an OSS CI config to catch this earlier on.

CI should already be building the code with `-Wall, -Wextra, -Werror`, so I'm not sure what is going on...

snadampal commented:
I've updated the commit to decorate the optional functions with unused, but i'm not able to reopen the PR.
https://github.com/snadampal/pytorch/commit/54507a76d23bb4185812c71dd4511f8f5f447f50

snadampal commented:
@seemethere  , @malfet  , please reopen the PR. thank you!

malfet commented:
@snadampal I can't reopen it, as you've deleted branch it backs. Please simply create new PR

snadampal commented:
The branch still exists, I have just rebased it to the latest master. if new PR is the only option, I will create.

snadampal commented:
Hi @malfet , here is the new PR: https://github.com/pytorch/pytorch/pull/95963
